File: app/src/main/java/com/example/recipeclipper/data/DefaultRecipeRepository.kt
package com.example.recipeclipper.data

import com.example.recipeclipper.data.local.CookStateJson
import com.example.recipeclipper.data.local.dao.RecipeDao
import com.example.recipeclipper.data.local.entity.newUid
import com.example.recipeclipper.data.model.ContentOrigin
import com.example.recipeclipper.data.model.CookProgress
import com.example.recipeclipper.data.model.ManualRecipe
import com.example.recipeclipper.data.model.RecipeDraft
import com.example.recipeclipper.data.model.ParseError
import com.example.recipeclipper.data.model.PlanDays
import com.example.recipeclipper.data.model.ParseResult
import com.example.recipeclipper.data.model.Recipe
import com.example.recipeclipper.data.model.RecipeSummary
import com.example.recipeclipper.data.model.StepAlarm
import com.example.recipeclipper.data.model.UrlCleaner
import com.example.recipeclipper.data.flags.Flag
import com.example.recipeclipper.data.flags.FeatureFlags
import com.example.recipeclipper.data.model.PageText
import com.example.recipeclipper.data.model.RecipeTextWindow
import com.example.recipeclipper.data.remote.BlogRecipeSource
import com.example.recipeclipper.data.remote.FetchedPage
import com.example.recipeclipper.data.remote.PageRecipe
import com.example.recipeclipper.data.remote.RecipeSource
import com.example.recipeclipper.data.remote.RenderedPageSource
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import javax.inject.Inject
import javax.inject.Singleton

/**
 * The real, Room-backed [RecipeRepository]. Bound to the interface with `@Binds`.
 *
 * Database failures never escape: each call runs through [ErrorLog.guard], is logged, and
 * degrades to what the contract already allows — `Error(SaveFailed)` from an import or a clip, null from
 * [open] and [delete], nothing from [setChecked], [setNotes], [setCookProgress],
 * [setServingsTarget] and [restore], none from [runningTimers], an empty list from a Flow —
 * rather than crashing `viewModelScope`. The iOS repository does the same.
 */
@Singleton
class DefaultRecipeRepository @Inject constructor(
    private val source: RecipeSource,
    private val recipeDao: RecipeDao,
    private val clock: Clock,
    private val log: ErrorLog,
    private val renderedPages: RenderedPageSource = RenderedPageSource.None,
    private val library: LibraryPolicy = LibraryPolicy.HistoryOnly,
    private val extractor: PageRecipeExtractor = PageRecipeExtractor.None,
    /** Null (tests that aren't about it) reads as every flag off. */
    private val flags: FeatureFlags? = null
) : RecipeRepository {

    override suspend fun importFromUrl(sharedUrl: String): ParseResult {
        // Saved under the cleaned link, so tracking tags can't make one recipe into two.
        val url = UrlCleaner.clean(sharedUrl)
        // The user's version is never refreshed by a re-share (#29): open it without a fetch.
        val usersVersion = log.guard("find user's version", null) {
            recipeDao.findByUrl(url)?.takeIf { !ContentOrigin.isSources(it.contentOrigin) }
        }
        if (usersVersion != null) {
            val viewedAt = clock.now()
            return log.guard("open user's version", ParseResult.Error(ParseError.SaveFailed)) {
                recipeDao.upsert(usersVersion.copy(lastViewedAt = viewedAt), library.limit, today = today(viewedAt))
                ParseResult.Success(usersVersion.copy(lastViewedAt = viewedAt).toDomain())
            }
        }
        val parsed = fetchWithFallbacks(url)
        val now = clock.now()
        return when (parsed) {
            is ParseResult.Success -> log.guard("import save", ParseResult.Error(ParseError.SaveFailed)) {
                save(parsed.recipe, now)
            }
            is ParseResult.Error -> {
                // Offline, blocked or anything else: a link opened once still opens.
                val cached = log.guard("offline fallback", null) {
                    recipeDao.findByUrl(url)?.also { recipeDao.touch(it.id, now) }
                } ?: return parsed
                ParseResult.Success(cached.copy(lastViewedAt = now).toDomain())
            }
        }
    }

    override suspend fun updateFromSource(id: Long): ParseResult {
        val existing = log.guard("updateFromSource find", null) { recipeDao.get(id) }
            ?: return ParseResult.Error(ParseError.NotSaved)
        if (!existing.toDomain().canUpdateFromSource) return ParseResult.Error(ParseError.NothingToShow)
        val parsed = fetchWithFallbacks(existing.sourceUrl)
        if (parsed !is ParseResult.Success) return parsed
        val now = clock.now()
        // Filed under the saved link, whatever the parse reports, so it lands on this row.
        val fresh = parsed.recipe.copy(sourceUrl = existing.sourceUrl).toEntity(now)
        return log.guard("updateFromSource save", ParseResult.Error(ParseError.SaveFailed)) {
            recipeDao.upsert(fresh, library.limit, replaceUsersVersion = true, today = today(now))
            recipeDao.get(id)?.let { ParseResult.Success(it.toDomain()) }
                ?: ParseResult.Error(ParseError.SaveFailed)
        }
    }

    override suspend fun saveEdit(id: Long, draft: RecipeDraft): Recipe? {
        if (!draft.isValid) return null
        return log.guard("saveEdit", null) {
            val existing = recipeDao.get(id)?.toDomain() ?: return@guard null
            val edited = draft.applyTo(existing).toEntity(existing.lastViewedAt)
            val saved = recipeDao.saveEdit(id, edited, existing.origin.afterEdit().name, clock.now())
            if (saved) recipeDao.get(id)?.toDomain() else null
        }
    }

    override suspend fun addManual(draft: RecipeDraft): Recipe? {
        if (!draft.isValid) return null
        val now = clock.now()
        val recipe = draft.applyTo(
            Recipe(
                name = "", image = null, ingredients = emptyList(), instructions = emptyList(),
                prepTime = null, cookTime = null, totalTime = null, yield = null,
                sourceUrl = ManualRecipe.newSourceUrl(newUid()),
                origin = ContentOrigin.MANUAL,
                editedAt = now
            )
        )
        return log.guard("addManual", null) {
            val id = recipeDao.upsert(recipe.toEntity(now), library.limit, today = today(now))
            if (id == RecipeDao.NOT_KEPT) recipe else recipeDao.get(id)?.toDomain()
        }
    }

    /**
     * [fetchWithOneRetry], then, only if that still ends [ParseError.Blocked] or
     * [ParseError.NoRecipeFound] (see [ParseError.triesRenderedPage]), one load of the page in an
     * off-screen browser, capped at [RENDER_TIMEOUT_MS], its HTML run through the same parsers.
     * Never after Offline or a timeout. A rendered page that still has no recipe, or that
     * doesn't load, leaves the direct fetch's cause standing: a block stays a block. Nothing
     * is shown for it; the import is just slower. Cancelling during it throws out of here
     * before anything is written.
     *
     * Last, only when the page loaded with no recipe data ([ParseError.NoRecipeFound]), the
     * on-device model may pick one out of its text ([extractFromPage]; the rendered page's
     * text if there is one).
     */
    private suspend fun fetchWithFallbacks(url: String): ParseResult {
        val fetched = fetchWithOneRetry(url)
        val result = fetched.result
        if (result !is ParseResult.Error || !result.error.triesRenderedPage) return result
        val html = withTimeoutOrNull(RENDER_TIMEOUT_MS) { renderedPages.render(url) }
        val rendered = html?.let { withContext(Dispatchers.Default) { BlogRecipeSource.parsePage(it, url) } }
        if (rendered != null && rendered.result is ParseResult.Success) return rendered.result
        if (result.error != ParseError.NoRecipeFound) return result
        val page = rendered?.page ?: fetched.page ?: return result
        return extractFromPage(page, url) ?: result
    }

    /**
     * A recipe the on-device model picked out of [page]'s text (#103), behind the
     * `llmExtraction` flag: the part of the page most likely to hold it ([RecipeTextWindow]),
     * then only what [PageRecipe.recipe] finds on the page as written. Null, and the page stays
     * [ParseError.NoRecipeFound] as before, on a phone or in a language the model can't read,
     * or when too little of what it picked is on the page.
     */
    private suspend fun extractFromPage(page: PageText, url: String): ParseResult.Success? {
        if (flags?.isOn(Flag.LLM_EXTRACTION) != true) return null
        val language = withContext(Dispatchers.Default) { PageRecipe.language(page) }
        val chars = extractor.windowChars(language) ?: return null
        val window = withContext(Dispatchers.Default) { RecipeTextWindow.window(page, chars) } ?: return null
        val picked = extractor.extract(window, language) ?: return null
        val recipe = withContext(Dispatchers.Default) { PageRecipe.recipe(window, picked, page, url) }
        return recipe?.let { ParseResult.Success(it) }
    }

    /**
     * Fetches, and if that fails in a way that often clears on its own (a block, or a network
     * failure that wasn't a timeout; see [ParseError.shouldAutoRetry]) waits [RETRY_PAUSE_MS]
     * and fetches once more. Offline fails straight away, a timeout isn't repeated (so a dead
     * Wi-Fi costs one timeout, not two), and a page that loaded with no recipe is never
     * retried. The pause is a plain coroutine [delay], so a test's virtual time skips it, and
     * cancelling the import during it throws out of here before anything is written.
     */
    private suspend fun fetchWithOneRetry(url: String): FetchedPage {
        val first = source.fetchPage(url)
        val error = (first.result as? ParseResult.Error)?.error
        if (error == null || !error.shouldAutoRetry) return first
        delay(RETRY_PAUSE_MS)
        return source.fetchPage(url)
    }

    override suspend fun saveClip(recipe: Recipe): ParseResult =
        log.guard("saveClip", ParseResult.Error(ParseError.SaveFailed)) {
            // CLIPPED, so a re-share opens the clip rather than fetching the page again (#29's
            // rule); a clip replaces whatever the row held, as "Update from source" does.
            val clip = recipe.copy(
                sourceUrl = UrlCleaner.clean(recipe.sourceUrl),
                origin = ContentOrigin.CLIPPED,
                editedAt = null
            )
            val now = clock.now()
            val id = recipeDao.upsert(clip.toEntity(now), library.limit, replaceUsersVersion = true, today = today(now))
            saved(id, clip)
        }

    override suspend fun keep(recipe: Recipe): ParseResult =
        log.guard("keep", ParseResult.Error(ParseError.SaveFailed)) { save(recipe, clock.now()) }

    /** Saves a parsed recipe as a view at [now], under the library's limit (#107). */
    private suspend fun save(recipe: Recipe, now: Long): ParseResult =
        saved(recipeDao.upsert(recipe.toEntity(now), library.limit, today = today(now)), recipe)

    /** The row [id] as saved, or [shown] not kept when the full library had no room. */
    private suspend fun saved(id: Long, shown: Recipe): ParseResult =
        if (id == RecipeDao.NOT_KEPT) {
            ParseResult.Success(shown.copy(id = 0), kept = false)
        } else {
            recipeDao.get(id)?.let { ParseResult.Success(it.toDomain()) }
                ?: ParseResult.Error(ParseError.SaveFailed)
        }

    override suspend fun open(id: Long): Recipe? = log.guard("open", null) {
        val entity = recipeDao.get(id) ?: return@guard null
        val now = clock.now()
        recipeDao.touch(id, now)
        entity.copy(lastViewedAt = now).toDomain()
    }

    override suspend fun setChecked(id: Long, checked: Set<Int>) =
        log.guard("setChecked", Unit) { recipeDao.setChecked(id, checked) }

    override suspend fun setNotes(id: Long, notes: String) =
        log.guard("setNotes", Unit) { recipeDao.setNotes(id, notes.takeIf { it.isNotBlank() }) }

    override suspend fun setCookProgress(id: Long, progress: CookProgress) =
        log.guard("setCookProgress", Unit) { recipeDao.setCookState(id, CookStateJson.encode(progress)) }

    override suspend fun setServingsTarget(id: Long, target: Int?) =
        log.guard("setServingsTarget", Unit) { recipeDao.setServingsTarget(id, target) }

    override suspend fun runningTimers(): List<StepAlarm> = log.guard("runningTimers", emptyList()) {
        recipeDao.cookStates().flatMap { row ->
            CookStateJson.decode(row.cookState).timers.mapNotNull { (step, timer) ->
                timer.endsAt?.let { StepAlarm(row.id, row.title, step, it) }
            }
        }
    }

    override suspend fun delete(id: Long): RecipeRepository.DeletedRecipe? = log.guard("delete", null) {
        val entity = recipeDao.get(id) ?: return@guard null
        val crossRefs = recipeDao.crossRefsFor(id)
        val planEntries = recipeDao.planEntriesFor(id)
        val menuEntries = recipeDao.menuEntriesFor(id)
        recipeDao.delete(id)
        RecipeRepository.DeletedRecipe(entity, crossRefs, planEntries, menuEntries)
    }

    override suspend fun restore(deleted: RecipeRepository.DeletedRecipe) =
        log.guard("restore", Unit) { recipeDao.restore(deleted.entity, deleted.crossRefs, deleted.planEntries, deleted.menuEntries) }

    override fun observeHistory(query: String): Flow<List<RecipeSummary>> =
        recipeDao.observeHistory(query).map { rows -> rows.map { it.toDomain() } }.orEmptyOnError(log, "observeHistory")

    override fun observeRecent(limit: Int): Flow<List<RecipeSummary>> =
        recipeDao.observeRecent(limit).map { rows -> rows.map { it.toDomain() } }.orEmptyOnError(log, "observeRecent")

    override fun observeCount(): Flow<Int> = recipeDao.observeCount().catch { e ->
        log.error("observeCount failed", e)
        emit(0)
    }

    /** Today on the user's calendar: a recipe planned for it or later is never culled (#49). */
    private fun today(now: Long): Long = PlanDays.today(now)

    companion object {
        /** How long to wait before the single automatic retry. Long enough for a momentary
         *  block to lift, short enough that the spinner doesn't feel stuck. */
        const val RETRY_PAUSE_MS = 2_000L

        /** The cap on the off-screen browser fallback, load and settle together. It comes on
         *  top of the direct fetch and its retry, so a page that never settles can't hold the
         *  spinner much longer than they did. */
        const val RENDER_TIMEOUT_MS = 20_000L
    }
}

File: ios/RecipeClipper/Data/Local/DefaultRecipeRepository.swift
import Combine
import Foundation

/// The real, SQLite-backed RecipeRepository (Android's DefaultRecipeRepository).
///
/// The contract's methods don't throw, so a database failure is logged and surfaced the way
/// the contract allows: `.error(.saveFailed)` from an import, nil from `open` / `delete`, and
/// nothing from `setChecked` / `setNotes` / `setCookProgress` / `setServingsTarget` / `restore`.
final class DefaultRecipeRepository: RecipeRepository {
    private let db: AppDatabase
    private let source: RecipeSource
    private let clock: Clock
    private let sleep: RetrySleep
    private let renderedPages: RenderedPageSource
    private let renderTimeout: Duration
    private let library: LibraryLimitSource
    private let photos: PhotoStore?
    private let extractor: PageRecipeExtractor
    /// The `llmExtraction` flag (#103), read at each import; the share extension leaves it off.
    private let extractionOn: () -> Bool

    /// How a failed fetch waits before its single retry. Injected so a test doesn't really wait;
    /// throws on cancellation, as `Task.sleep` does.
    typealias RetrySleep = (Duration) async throws -> Void

    /// How long to wait before the single automatic retry: long enough for a momentary block to
    /// lift, short enough that the spinner doesn't feel stuck. Android's RETRY_PAUSE_MS.
    static let retryPause: Duration = .seconds(2)

    /// The cap on the off-screen browser fallback, load and settle together. It comes on top of
    /// the direct fetch and its retry, so a page that never settles can't hold the spinner much
    /// longer than they did. Android's RENDER_TIMEOUT_MS.
    static let renderTimeout: Duration = .seconds(20)

    init(
        db: AppDatabase,
        source: RecipeSource,
        clock: Clock,
        sleep: @escaping RetrySleep = { try await Task.sleep(for: $0) },
        renderedPages: RenderedPageSource = NoRenderedPageSource(),
        renderTimeout: Duration = DefaultRecipeRepository.renderTimeout,
        library: LibraryLimitSource = FixedLibraryLimit(),
        extractor: PageRecipeExtractor = NoPageRecipeExtractor(),
        extractionOn: @escaping () -> Bool = { false },
        photos: PhotoStore? = nil
    ) {
        self.photos = photos
        self.library = library
        self.extractor = extractor
        self.extractionOn = extractionOn
        self.db = db
        self.source = source
        self.clock = clock
        self.sleep = sleep
        self.renderedPages = renderedPages
        self.renderTimeout = renderTimeout
    }

    func importFromUrl(_ sharedUrl: String, renderedPage: String?) async -> ParseResult {
        // Saved under the cleaned link, so tracking tags can't make one recipe into two.
        let url = UrlCleaner.clean(sharedUrl)

        // The user's version is never refreshed by a re-share (#29): open it without a fetch.
        do {
            let viewedAt = clock.now()
            let limit = library.current()
            let usersVersion = try await db.write { conn -> RecipeRecord? in
                let dao = RecipeDao(db: conn)
                guard var row = try dao.findByUrl(url), !ContentOrigin.isSources(row.contentOrigin)
                else { return nil }
                row.lastViewedAt = viewedAt
                _ = try dao.upsert(row, limit: limit, today: PlanDays.today(millis: viewedAt)) // a view, and the cull
                return row
            }
            if let usersVersion { return .success(usersVersion.toDomain()) }
        } catch {
            dataLog.error("find user's version failed: \(String(describing: error), privacy: .public)")
        }

        let parsed = await fetchWithFallbacks(url, renderedPage: renderedPage)

        // RecipeSource.fetch can't throw, so cancellation can't propagate as it does on
        // Android. Match Android's outcome instead: a cancelled import writes nothing (not
        // even a touch). The cancelled caller discards whatever is returned.
        if Task.isCancelled { return parsed }
        let now = clock.now()

        switch parsed {
        case .notKept:
            return parsed // a fetch never answers this: only a save does
        case .success(var recipe):
            // Pin the key to the cleaned link we looked up by, so the offline fallback below
            // always finds what was saved here. (The parser already sets this; belt and braces.)
            recipe.sourceUrl = url
            return await save(recipe, now: now, what: "import save")

        case .error:
            // Offline, blocked or anything else: anything opened once still opens.
            let cached = try? await db.write { conn -> RecipeRecord? in
                let dao = RecipeDao(db: conn)
                guard var row = try dao.findByUrl(url) else { return nil }
                try dao.touch(row.id, now: now)
                row.lastViewedAt = now
                return row
            }
            guard let cached = cached ?? nil else { return parsed }
            return .success(cached.toDomain())
        }
    }

    /// With a rendered page already in hand (Safari's share extension preprocessing, #35),
    /// parses it directly and skips the fetch entirely. Only when that page holds no recipe
    /// does the ordinary fetch — with its retry and rendered-browser fallback — run, exactly as
    /// if `renderedPage` had been nil.
    ///
    /// Otherwise: fetches, retries once after a pause if that often clears on its own, then
    /// loads the page once in an off-screen browser if it is still blocked or has no recipe
    /// data.
    private func fetchWithFallbacks(_ url: String, renderedPage: String?) async -> ParseResult {
        if let renderedPage {
            let fromPage = BlogRecipeSource.parse(html: renderedPage, url: url)
            if case .success = fromPage { return fromPage }
        }

        var fetched = await source.fetchPage(url: url)

        // A block or a network blip often clears on its own: wait, then fetch once more. Offline
        // fails straight away, a timeout isn't repeated (a dead Wi-Fi costs one timeout, not
        // two), and a page that loaded with no recipe is never retried.
        if case .error(let error) = fetched.result, error.shouldAutoRetry, !Task.isCancelled {
            do {
                try await sleep(Self.retryPause)
            } catch {
                return fetched.result // cancelled during the pause: the caller writes nothing
            }
            if !Task.isCancelled { fetched = await source.fetchPage(url: url) }
        }
        let parsed = fetched.result

        // Still blocked, or a page with no recipe data: load it once in an off-screen browser
        // and run what it renders through the same parsers. Never after offline or a timeout.
        // A rendered page with no recipe, or one that doesn't load, leaves the cause standing.
        guard case .error(let error) = parsed, error.triesRenderedPage, !Task.isCancelled else { return parsed }
        var rendered: FetchedPage?
        if let html = await renderCapped(url), !Task.isCancelled {
            rendered = BlogRecipeSource.parsePage(html: html, url: url)
            if let rendered, case .success = rendered.result { return rendered.result }
        }
        // Last, only for a page that loaded with no recipe data: the on-device model may pick
        // one out of its text (the rendered page's if there is one; #103).
        guard error == .noRecipeFound, !Task.isCancelled,
              let page = rendered?.page ?? fetched.page ?? renderedPage.map({ PageTextReader.read(html: $0, url: url) })
        else { return parsed }
        return await extractFromPage(page, url: url) ?? parsed
    }

    /// A recipe the on-device model picked out of `page`'s text (#103), behind the
    /// `llmExtraction` flag: the part most likely to hold it (`RecipeTextWindow`), then only what
    /// `PageRecipe.recipe` finds on the page as written. Nil, and the page stays
    /// `.noRecipeFound`, on a phone or in a language the model can't read, or when too little of
    /// what it picked is on the page. Android's `extractFromPage`.
    private func extractFromPage(_ page: PageText, url: String) async -> ParseResult? {
        guard extractionOn() else { return nil }
        let language = PageRecipe.language(page)
        guard let chars = await extractor.windowChars(language: language),
              let window = RecipeTextWindow.window(page, maxChars: chars),
              let picked = await extractor.extract(window, language: language), !Task.isCancelled,
              let recipe = PageRecipe.recipe(window: window, picked: picked, page: page, url: url)
        else { return nil }
        return .success(recipe)
    }

    func updateFromSource(id: Int64) async -> ParseResult {
        guard let existing = (try? await db.read { conn in try RecipeDao(db: conn).get(id) }) ?? nil
        else { return .error(.notSaved) }
        guard existing.toDomain().canUpdateFromSource else { return .error(.nothingToShow) }
        let parsed = await fetchWithFallbacks(existing.sourceUrl, renderedPage: nil)
        if Task.isCancelled { return parsed }
        guard case .success(var recipe) = parsed else { return parsed }
        // Filed under the saved link, whatever the parse reports, so it lands on this row.
        recipe.sourceUrl = existing.sourceUrl
        let fresh = recipe.toRecord(viewedAt: clock.now())
        let limit = library.current()
        do {
            let saved = try await db.write { conn -> RecipeRecord? in
                let dao = RecipeDao(db: conn)
                _ = try dao.upsert(
                    fresh, limit: limit, replaceUsersVersion: true,
                    today: PlanDays.today(millis: fresh.lastViewedAt)
                )
                return try dao.get(id)
            }
            guard let saved else { return .error(.saveFailed) }
            return .success(saved.toDomain())
        } catch {
            dataLog.error("updateFromSource save failed: \(String(describing: error), privacy: .public)")
            return .error(.saveFailed)
        }
    }

    func saveEdit(id: Int64, draft: RecipeDraft) async -> Recipe? {
        guard draft.isValid else { return nil }
        let now = clock.now()
        do {
            return try await db.write { conn -> Recipe? in
                let dao = RecipeDao(db: conn)
                guard let existing = try dao.get(id)?.toDomain() else { return nil }
                let edited = draft.apply(to: existing).toRecord(viewedAt: existing.lastViewedAt)
                guard try dao.saveEdit(id, edited: edited, origin: existing.origin.afterEdit().rawValue, editedAt: now)
                else { return nil }
                return try dao.get(id)?.toDomain()
            }
        } catch {
            dataLog.error("saveEdit failed: \(String(describing: error), privacy: .public)")
            return nil
        }
    }

    func addManual(draft: RecipeDraft) async -> Recipe? {
        guard draft.isValid else { return nil }
        let now = clock.now()
        var recipe = draft.apply(to: Recipe(
            name: "", image: nil, ingredients: [], instructions: [],
            prepTime: nil, cookTime: nil, totalTime: nil, yield: nil,
            sourceUrl: ManualRecipe.newSourceUrl(newUid())
        ))
        recipe.origin = .manual
        recipe.editedAt = now
        let record = recipe.toRecord(viewedAt: now)
        let limit = library.current()
        do {
            return try await db.write { conn -> Recipe? in
                let dao = RecipeDao(db: conn)
                let id = try dao.upsert(record, limit: limit, today: PlanDays.today(millis: now))
                if id == RecipeDao.notKept { return recipe }
                return try dao.get(id)?.toDomain()
            }
        } catch {
            dataLog.error("addManual failed: \(String(describing: error), privacy: .public)")
            return nil
        }
    }

    /// The rendered page, or nil once `renderTimeout` runs out (which cancels the render).
    private func renderCapped(_ url: String) async -> String? {
        let renderedPages = renderedPages
        let renderTimeout = renderTimeout
        return await withTaskGroup(of: String?.self) { group in
            group.addTask { await renderedPages.render(url: url) }
            group.addTask {
                try? await Task.sleep(for: renderTimeout)
                return nil
            }
            let first = await group.next() ?? nil
            group.cancelAll()
            return first
        }
    }

    func saveClip(_ recipe: Recipe) async -> ParseResult {
        // CLIPPED, so a re-share opens the clip rather than fetching the page again (#29's
        // rule); a clip replaces whatever the row held, as "Update from source" does.
        var clip = recipe
        clip.sourceUrl = UrlCleaner.clean(recipe.sourceUrl)
        clip.origin = .clipped
        clip.editedAt = nil
        return await save(clip, now: clock.now(), replaceUsersVersion: true, what: "clip save")
    }

    func keep(_ recipe: Recipe) async -> ParseResult {
        await save(recipe, now: clock.now(), what: "keep")
    }

    /// Saves `recipe` as a view at `now`, under the library's limit (#107): the saved row, or
    /// `.notKept` when the full library had no room.
    private func save(_ recipe: Recipe, now: Int64, replaceUsersVersion: Bool = false, what: String) async -> ParseResult {
        let limit = library.current()
        do {
            let saved = try await db.write { conn -> RecipeRecord?? in
                let dao = RecipeDao(db: conn)
                let id = try dao.upsert(
                    recipe.toRecord(viewedAt: now), limit: limit, replaceUsersVersion: replaceUsersVersion,
                    today: PlanDays.today(millis: now)
                )
                if id == RecipeDao.notKept { return .some(nil) }
                return .some(try dao.get(id))
            }
            switch saved {
            case .some(.some(let row)): return .success(row.toDomain())
            case .some(.none):
                var shown = recipe
                shown.id = 0
                return .notKept(shown)
            case .none: return .error(.saveFailed)
            }
        } catch {
            dataLog.error("\(what, privacy: .public) failed: \(String(describing: error), privacy: .public)")
            return .error(.saveFailed)
        }
    }

    func open(id: Int64) async -> Recipe? {
        let now = clock.now()
        do {
            let row = try await db.write { conn -> RecipeRecord? in
                let dao = RecipeDao(db: conn)
                guard var row = try dao.get(id) else { return nil }
                try dao.touch(id, now: now)
                row.lastViewedAt = now
                return row
            }
            return row?.toDomain()
        } catch {
            dataLog.error("open failed: \(String(describing: error), privacy: .public)")
            return nil
        }
    }

    func setChecked(id: Int64, checked: Set<Int>) async {
        do {
            try await db.write { conn in try RecipeDao(db: conn).setChecked(id, checked: checked) }
        } catch {
            dataLog.error("setChecked failed: \(String(describing: error), privacy: .public)")
        }
    }

    func setNotes(id: Int64, notes: String) async {
        let stored: String? = notes.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : notes
        do {
            try await db.write { conn in try RecipeDao(db: conn).setNotes(id, notes: stored) }
        } catch {
            dataLog.error("setNotes failed: \(String(describing: error), privacy: .public)")
        }
    }

    func setCookProgress(id: Int64, progress: CookProgress) async {
        let json = CookStateJSON.encode(progress)
        do {
            try await db.write { conn in try RecipeDao(db: conn).setCookState(id, cookState: json) }
        } catch {
            dataLog.error("setCookProgress failed: \(String(describing: error), privacy: .public)")
        }
    }

    func setServingsTarget(id: Int64, target: Int?) async {
        do {
            try await db.write { conn in try RecipeDao(db: conn).setServingsTarget(id, target: target) }
        } catch {
            dataLog.error("setServingsTarget failed: \(String(describing: error), privacy: .public)")
        }
    }

    func delete(id: Int64) async -> DeletedRecipe? {
        do {
            return try await db.write { conn -> DeletedRecipe? in
                let dao = RecipeDao(db: conn)
                guard let row = try dao.get(id) else { return nil }
                // Read before deleting: the cascade takes them with the row.
                let memberships = try dao.crossRefsFor(id)
                let planEntries = try dao.planEntriesFor(id)
                let menuEntries = try dao.menuEntriesFor(id)
                let cookedPhotos = try CookedPhotoDao(db: conn).photosFor(id)
                try dao.delete(id)
                return DeletedRecipe(
                    recipe: row.toDomain(), memberships: memberships, uid: row.uid, planEntries: planEntries,
                    menuEntries: menuEntries, cookedPhotos: cookedPhotos
                )
            }
        } catch {
            dataLog.error("delete failed: \(String(describing: error), privacy: .public)")
            return nil
        }
    }

    func restore(_ deleted: DeletedRecipe) async {
        do {
            var record = deleted.recipe.toRecord(viewedAt: deleted.recipe.lastViewedAt)
            if let uid = deleted.uid { record.uid = uid }
            try await db.write { [record] conn in
                try RecipeDao(db: conn).restore(record, crossRefs: deleted.memberships, planEntries: deleted.planEntries,
                    menuEntries: deleted.menuEntries, cookedPhotos: deleted.cookedPhotos
                )
            }
        } catch {
            dataLog.error("restore failed: \(String(describing: error), privacy: .public)")
        }
    }

    // The photos belong to the recipe (#116): once its delete stands, their files go too.
    func forget(_ deleted: DeletedRecipe) async {
        guard let photos, !deleted.cookedPhotos.isEmpty else { return }
        await photos.delete(deleted.cookedPhotos.map(\.fileName))
    }

    func observeHistory(query: String) -> AnyPublisher<[RecipeSummary], Never> {
        db.observe { conn in try RecipeDao(db: conn).history(query: query).map { $0.toDomain() } }
    }

    func observeRecent(limit: Int) -> AnyPublisher<[RecipeSummary], Never> {
        db.observe { conn in try RecipeDao(db: conn).recent(limit: limit).map { $0.toDomain() } }
    }

    /// Every recipe but the tour's sample (#151), which never takes a free-tier place.
    func observeCount() -> AnyPublisher<Int, Never> {
        db.observe { conn in try RecipeDao(db: conn).count() }
    }

    func sampleId() async -> Int64? {
        do {
            return try await db.read { conn in try RecipeDao(db: conn).findByUrl(SampleRecipe.sourceUrl)?.id }
        } catch {
            dataLog.error("sampleId failed: \(String(describing: error), privacy: .public)")
            return nil
        }
    }

    func addSample(_ recipe: Recipe) async -> Int64? {
        var sample = recipe
        sample.sourceUrl = SampleRecipe.sourceUrl
        sample.origin = .manual
        let record = sample.toRecord(viewedAt: clock.now())
        do {
            // Unlimited: the sample counts toward no limit, so adding it never removes a recipe.
            let id = try await db.write { conn in try RecipeDao(db: conn).upsert(record, limit: .unlimited) }
            return id == RecipeDao.notKept ? nil : id
        } catch {
            dataLog.error("addSample failed: \(String(describing: error), privacy: .public)")
            return nil
        }
    }
}
